### PR TITLE
External sampling mccfr in python

### PR DESCRIPTION
**Pull request by colibri17**
I implemented the external version of MCCFR in Python (`external_sampling_mccfr.py`). I tried to mirror the C++ version (`external_sampling_mccfr.cc`) as much as possible, taking the scheleton from the outcome version in Python (`outcome_sampling_mccfr.py`). Some code is duplicated with respect to it and maybe a super-class should be created.

I also created a module for tests (`external_sampling_mccfr_test.py`) where I taken some of the threshold values from the C++ version. I added the remaining values without any particular logic, but just using small thresholds similar to the known cases.

Finally, I changed the `mccfr_example.py` by allowing to use the new external sampling.

### CODE REVIEW

**Comment by colibri17** [comment]
@googlebot I fixed it.

**Comment by colibri17** [comment]
@googlebot I fixed it.

**Comment by lanctot** [comment]
Cool, thanks @colibri17 ! I will take a look early this week.

**Comment by colibri17** [comment]
You're welcome! That's great @lanctot thank you.

File: open_spiel/python/algorithms/external_sampling_mccfr.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+_REGRET_INDEX = 0
+_AVG_POLICY_INDEX = 1
+
+
+class ExternalSamplingSolver(object):
+    """
+    An implementation of external sampling MCCFR
+    """
+
+    def __init__(self, game, average_type='k_simple'):
```
Comment (lanctot) [inline open_spiel/python/algorithms/external_sampling_mccfr.py:34]: Can we make the average_type an enum defined at the top of this file?

File: open_spiel/python/algorithms/external_sampling_mccfr_test.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+import numpy as np
+from open_spiel.python import policy
+from open_spiel.python.algorithms import exploitability
+from open_spiel.python.algorithms import external_sampling_mccfr
+import pyspiel
+
+SEED = 39823987
+
+
+class OutcomeSamplingMCCFRTest(absltest.TestCase):
```
Comment (lanctot) [inline open_spiel/python/algorithms/external_sampling_mccfr_test.py:31]: Rename to ExternalSamplingMCCFRTest

File: open_spiel/python/algorithms/external_sampling_mccfr_test.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+    def test_outcome_sampling_leduc_2p_simple(self):
+        np.random.seed(SEED)
+        game = pyspiel.load_game("leduc_poker")
+        os_solver = external_sampling_mccfr.ExternalSamplingSolver(game, 'k_simple')
```
Comment (lanctot) [inline open_spiel/python/algorithms/external_sampling_mccfr_test.py:36]: es_solver, same below

**Review by lanctot — CHANGES_REQUESTED** [review]

**Comment by lanctot** [comment]
Overall, very nice! I only have a few minor comments. I'll import it once you fix these.

File: open_spiel/python/algorithms/external_sampling_mccfr.py (hunk 1 above)
Comment (colibri17) [reply open_spiel/python/algorithms/external_sampling_mccfr.py:34]: I am not sure I get it. Do you mean define at the top of this file something like:
```
from enum import Enum
average_type = Enum('average_type', 'k_simple k_full')
```
And then having it called in the constructor as default parameter by using:
`def __init__(self, game, average_type=average_type.k_simple.name)`?

File: open_spiel/python/algorithms/external_sampling_mccfr_test.py (hunk 3 above)
Comment (colibri17) [reply open_spiel/python/algorithms/external_sampling_mccfr_test.py:36]: Yes, good point. Forgot to change the names!

File: open_spiel/python/algorithms/external_sampling_mccfr_test.py (hunk 2 above)
Comment (colibri17) [reply open_spiel/python/algorithms/external_sampling_mccfr_test.py:31]: Yes, good point. Forgot to change the method names!

File: open_spiel/python/algorithms/external_sampling_mccfr.py (hunk 1 above)
Comment (lanctot) [reply open_spiel/python/algorithms/external_sampling_mccfr.py:34]: Yes but no worries I can change it upon import.

Comment (colibri17) [reply open_spiel/python/algorithms/external_sampling_mccfr.py:34]: Ok, will let it as it is then. Thanks

**Comment by colibri17** [comment]
I did another push with the fixes. Let me know if everything is ok

**Review by lanctot — APPROVED** [review]

**Comment by lanctot** [comment]
> I did another push with the fixes. Let me know if everything is ok

Looking good, importing now.

**Pull request merged**